File: core/src/test/kotlin/com/auto_switch_ime/core/ime/ImeGatewayTest.kt
package com.auto_switch_ime.core.ime

import com.auto_switch_ime.core.ImeAsciiModeSwitcher
import com.auto_switch_ime.core.ImeProvider
import com.auto_switch_ime.core.ImeStateSource
import com.auto_switch_ime.core.ImeType
import com.auto_switch_ime.core.ime.system.SystemImeProvider
import com.auto_switch_ime.core.ime.system.SystemImeProviderRegistry
import com.auto_switch_ime.core.ime.system.SystemType
import com.auto_switch_ime.core.util.Logger
import kotlin.coroutines.Continuation
import kotlin.coroutines.EmptyCoroutineContext
import kotlin.coroutines.startCoroutine
import org.junit.jupiter.api.Assertions.assertEquals
import org.junit.jupiter.api.Assertions.assertFalse
import org.junit.jupiter.api.Assertions.assertThrows
import org.junit.jupiter.api.Assertions.assertTrue
import org.junit.jupiter.api.Test

class ImeGatewayTest {
    @Test
    fun `system state is the default for capabilities not supplied by input method`() {
        val system = FakeSystem(ascii = true, caps = true, composing = false)
        val gateway = ImeGateway(FakeProvider(), system, FakeLogger)

        assertEquals(
            com.auto_switch_ime.core.ImeState(true, true, false),
            gateway.getCurrentState()
        )
    }

    @Test
    fun `input method state overrides only fields it supplies`() {
        val source = object : ImeStateSource {
            override fun readAsciiMode(): Boolean = false
        }
        val system = FakeSystem(ascii = true, caps = true, composing = true)
        val gateway = ImeGateway(FakeProvider(stateSource = source), system, FakeLogger)

        assertEquals(
            com.auto_switch_ime.core.ImeState(false, true, true),
            gateway.getCurrentState()
        )
    }

    @Test
    fun `observed CapsLock state overrides background system reading`() {
        val system = FakeSystem(ascii = true, caps = true)
        val gateway = ImeGateway(FakeProvider(), system, FakeLogger)

        gateway.refreshState(observedCapsLock = false)

        assertFalse(gateway.getTrackedState().isCapsLock)
    }

    @Test
    fun `missing input method switcher falls back to system switcher`() {
        val system = FakeSystem(ascii = false)
        val gateway = ImeGateway(FakeProvider(), system, FakeLogger)
        gateway.refreshState()

        runSuspend { gateway.setAsciiMode(true) }

        assertEquals(1, system.asciiSwitchCount)
        assertTrue(gateway.getTrackedState().isAsciiMode)
    }

    @Test
    fun `input method switcher takes precedence over system switcher`() {
        val switcher = FakeAsciiSwitcher(succeeds = true)
        val system = FakeSystem(ascii = false)
        val gateway = ImeGateway(FakeProvider(switcher = switcher), system, FakeLogger)
        gateway.refreshState()

        runSuspend { gateway.setAsciiMode(true) }

        assertEquals(1, switcher.callCount)
        assertEquals(0, system.asciiSwitchCount)
    }

    @Test
    fun `forced ASCII mode invokes input method switcher when system state is already ASCII`() {
        val switcher = FakeAsciiSwitcher(succeeds = true)
        val system = FakeSystem(ascii = true)
        val gateway = ImeGateway(FakeProvider(switcher = switcher), system, FakeLogger)
        gateway.refreshState()

        runSuspend { gateway.setAsciiMode(true, forceAsciiMode = true) }

        assertEquals(1, switcher.callCount)
        assertEquals(0, system.asciiSwitchCount)
    }

    @Test
    fun `forced ASCII mode disables manual CapsLock before switching input method`() {
        val switcher = FakeAsciiSwitcher(succeeds = true)
        val system = FakeSystem(ascii = true, caps = true)
        val gateway = ImeGateway(FakeProvider(switcher = switcher), system, FakeLogger)
        gateway.refreshState()

        runSuspend { gateway.setAsciiMode(true, forceAsciiMode = true) }

        assertFalse(gateway.getTrackedState().isCapsLock)
        assertEquals(1, system.capsSwitchCount)
        assertEquals(1, switcher.callCount)
    }

    @Test
    fun `first input method action runs when provider state has not been observed`() {
        val switcher = FakeAsciiSwitcher(succeeds = true)
        val system = FakeSystem(ascii = true)
        val gateway = ImeGateway(FakeProvider(switcher = switcher), system, FakeLogger)
        gateway.refreshState()

        runSuspend { gateway.setAsciiMode(true) }

        assertEquals(1, switcher.callCount)
        assertTrue(gateway.getTrackedState().isAsciiMode)
    }

    @Test
    fun `input method state source remains authoritative when system state disagrees`() {
        var providerAscii = true
        val source = object : ImeStateSource {
            override fun readAsciiMode(): Boolean = providerAscii
        }
        val switcher = object : ImeAsciiModeSwitcher {
            var callCount = 0

            override suspend fun switchAsciiMode(
                ascii: Boolean,
                shouldContinue: () -> Boolean
            ): Boolean {
                callCount++
                if (!shouldContinue()) return false
                providerAscii = ascii
                return true
            }
        }
        val system = FakeSystem(ascii = true)
        val gateway = ImeGateway(
            FakeProvider(stateSource = source, switcher = switcher),
            system,
            FakeLogger
        )
        gateway.refreshState()

        runSuspend { gateway.setAsciiMode(false) }
        gateway.refreshState()

        assertFalse(gateway.getTrackedState().isAsciiMode)
        assertEquals(1, switcher.callCount)
        assertEquals(0, system.asciiSwitchCount)
    }

    @Test
    fun `successful switch command does not replace observed input method state`() {
        val source = object : ImeStateSource {
            override fun readAsciiMode(): Boolean = false
        }
        val switcher = FakeAsciiSwitcher(succeeds = true)
        val gateway = ImeGateway(
            FakeProvider(stateSource = source, switcher = switcher),
            FakeSystem(ascii = true),
            FakeLogger
        )
        gateway.refreshState()

        runSuspend { gateway.setAsciiMode(true) }

        assertFalse(gateway.getTrackedState().isAsciiMode)
        assertEquals(1, switcher.callCount)
    }

    @Test
    fun `system change updates tracked input method state`() {
        val switcher = FakeAsciiSwitcher(succeeds = true)
        val system = FakeSystem(ascii = true)
        val gateway = ImeGateway(FakeProvider(switcher = switcher), system, FakeLogger)
        gateway.refreshState()
        runSuspend { gateway.setAsciiMode(true) }

        system.setAsciiState(false)
        gateway.refreshState()

        assertFalse(gateway.getTrackedState().isAsciiMode)
    }

    @Test
    fun `failed input method switch does not fall back to system switcher`() {
        val switcher = FakeAsciiSwitcher(succeeds = false)
        val system = FakeSystem(ascii = false)
        val gateway = ImeGateway(FakeProvider(switcher = switcher), system, FakeLogger)
        gateway.refreshState()

        runSuspend { gateway.setAsciiMode(true) }

        assertEquals(1, switcher.callCount)
        assertEquals(0, system.asciiSwitchCount)
        assertFalse(gateway.getTrackedState().isAsciiMode)
    }

    @Test
    fun `declared unavailable state source suspends actions until it recovers`() {
        var available = false
        val source = object : ImeStateSource {
            override fun isAvailable(): Boolean = available
            override fun readAsciiMode(): Boolean = false
            override fun readComposing(): Boolean = false
        }
        val switcher = FakeAsciiSwitcher(succeeds = true)
        val gateway = ImeGateway(
            FakeProvider(stateSource = source, switcher = switcher),
            FakeSystem(ascii = true),
            FakeLogger
        )

        gateway.start()
        runSuspend { gateway.setAsciiMode(false) }

        assertFalse(gateway.isStateSourceAvailable())
        assertEquals(0, switcher.callCount)

        available = true
        gateway.refreshState()
        runSuspend { gateway.setAsciiMode(true) }

        assertTrue(gateway.isStateSourceAvailable())
        assertEquals(1, switcher.callCount)
    }

    @Test
    fun `system providers are selected through registry`() {
        val registry = SystemImeProviderRegistry()
        val windows = FakeSystem()
        registry.register(SystemType.WINDOWS) { windows }

        assertEquals(windows, registry.create(SystemType.WINDOWS))
        assertThrows(IllegalStateException::class.java) {
            registry.create(SystemType.LINUX)
        }
        assertEquals(SystemType.WINDOWS, SystemType.current("Windows 11"))
        assertEquals(SystemType.MACOS, SystemType.current("Mac OS X"))
        assertEquals(SystemType.LINUX, SystemType.current("Linux"))
    }

    private class FakeProvider(
        override val stateSource: ImeStateSource? = null,
        switcher: ImeAsciiModeSwitcher? = null
    ) : ImeProvider {
        override val type = ImeType.RIME
        override val name = "fake"
        override val asciiModeSwitcher = switcher
        override fun start() = Unit
        override fun dispose() = Unit
    }

    private class FakeAsciiSwitcher(private val succeeds: Boolean) : ImeAsciiModeSwitcher {
        var callCount = 0

        override suspend fun switchAsciiMode(
            ascii: Boolean,
            shouldContinue: () -> Boolean
        ): Boolean {
            callCount++
            return succeeds && shouldContinue()
        }
    }

    private class FakeSystem(
        private var ascii: Boolean = true,
        private var caps: Boolean = false,
        private var composing: Boolean = false
    ) : SystemImeProvider {
        override val type = SystemType.WINDOWS
        var asciiSwitchCount = 0
        var capsSwitchCount = 0

        override fun readAsciiMode(): Boolean = ascii
        override fun readCapsLock(): Boolean = caps
        override fun readComposing(): Boolean = composing

        fun setAsciiState(ascii: Boolean) {
            this.ascii = ascii
        }

        override suspend fun switchAsciiMode(
            ascii: Boolean,
            shouldContinue: () -> Boolean
        ): Boolean {
            asciiSwitchCount++
            if (!shouldContinue()) return false
            this.ascii = ascii
            return true
        }

        override suspend fun switchCapsLock(
            enabled: Boolean,
            shouldContinue: () -> Boolean
        ): Boolean {
            capsSwitchCount++
            if (!shouldContinue()) return false
            caps = enabled
            return true
        }
    }

    private object FakeLogger : Logger {
        override fun debug(msg: String) = Unit
        override fun info(msg: String) = Unit
        override fun warn(msg: String, e: Throwable?) = Unit
        override fun error(msg: String, e: Throwable?) = Unit
    }

    private fun runSuspend(block: suspend () -> Unit) {
        var outcome: Result<Unit>? = null
        block.startCoroutine(object : Continuation<Unit> {
            override val context = EmptyCoroutineContext

            override fun resumeWith(result: Result<Unit>) {
                outcome = result
            }
        })
        outcome?.getOrThrow() ?: error("Suspended test did not complete synchronously")
    }
}

File: vscode/test/ime-gateway.test.js
const assert = require('node:assert/strict');
const test = require('node:test');

const { ImeGateway } = require('../out/ime/ImeGateway');
const {
  currentSystemType,
  SystemImeProviderRegistry,
  SystemType,
} = require('../out/ime/system/SystemImeProvider');
const { ImeType } = require('../out/core/types');

const logger = {
  info() {},
  warn() {},
  debug() {},
  error() {},
};

function provider(overrides = {}) {
  return {
    type: ImeType.RIME,
    name: 'fake',
    start() {},
    dispose() {},
    ...overrides,
  };
}

function system(initial = {}) {
  const state = {
    ascii: initial.ascii ?? true,
    caps: initial.caps ?? false,
    composing: initial.composing ?? false,
  };
  return {
    type: SystemType.WINDOWS,
    asciiSwitchCount: 0,
    readAsciiMode: () => state.ascii,
    readCapsLock: () => state.caps,
    readComposing: () => state.composing,
    setAsciiState: (ascii) => { state.ascii = ascii; },
    async switchAsciiMode(ascii, shouldContinue) {
      this.asciiSwitchCount++;
      if (!shouldContinue()) return false;
      state.ascii = ascii;
      return true;
    },
    async switchCapsLock(enabled, shouldContinue) {
      if (!shouldContinue()) return false;
      state.caps = enabled;
      return true;
    },
  };
}

test('system state supplies capabilities missing from input method provider', () => {
  const ime = new ImeGateway(provider(), system({ ascii: true, caps: true }), logger);

  assert.deepEqual(ime.getCurrentState(), {
    isAsciiMode: true,
    isCapsLock: true,
    isComposing: false,
  });
});

test('input method state overrides only fields it supplies', () => {
  const imeProvider = provider({
    stateSource: { readAsciiMode: () => false },
  });
  const ime = new ImeGateway(
    imeProvider,
    system({ ascii: true, caps: true, composing: true }),
    logger,
  );

  assert.deepEqual(ime.getCurrentState(), {
    isAsciiMode: false,
    isCapsLock: true,
    isComposing: true,
  });
});

test('missing input method switcher falls back to system switcher', async () => {
  const systemProvider = system({ ascii: false });
  const ime = new ImeGateway(provider(), systemProvider, logger);
  ime.refreshState();

  await ime.setAsciiMode(true);

  assert.equal(systemProvider.asciiSwitchCount, 1);
  assert.equal(ime.getTrackedState().isAsciiMode, true);
});

test('input method switcher takes precedence and failure does not fall back', async () => {
  let providerSwitchCount = 0;
  const systemProvider = system({ ascii: false });
  const imeProvider = provider({
    asciiModeSwitcher: {
      async switchAsciiMode() {
        providerSwitchCount++;
        return false;
      },
    },
  });
  const ime = new ImeGateway(imeProvider, systemProvider, logger);
  ime.refreshState();

  await ime.setAsciiMode(true);

  assert.equal(providerSwitchCount, 1);
  assert.equal(systemProvider.asciiSwitchCount, 0);
  assert.equal(ime.getTrackedState().isAsciiMode, false);
});

test('forced ASCII mode invokes input method switcher when system state is already ASCII', async () => {
  let providerSwitchCount = 0;
  const systemProvider = system({ ascii: true });
  const imeProvider = provider({
    asciiModeSwitcher: {
      async switchAsciiMode() {
        providerSwitchCount++;
        return true;
      },
    },
  });
  const ime = new ImeGateway(imeProvider, systemProvider, logger);
  ime.refreshState();

  await ime.setAsciiMode(true, () => true, true);

  assert.equal(providerSwitchCount, 1);
  assert.equal(systemProvider.asciiSwitchCount, 0);
});

test('first input method action runs when provider state has not been observed', async () => {
  let providerSwitchCount = 0;
  const systemProvider = system({ ascii: true });
  const imeProvider = provider({
    asciiModeSwitcher: {
      async switchAsciiMode() {
        providerSwitchCount++;
        return true;
      },
    },
  });
  const ime = new ImeGateway(imeProvider, systemProvider, logger);
  ime.refreshState();

  await ime.setAsciiMode(true);

  assert.equal(providerSwitchCount, 1);
  assert.equal(ime.getTrackedState().isAsciiMode, true);
});

test('input method state source remains authoritative when system state disagrees', async () => {
  let providerSwitchCount = 0;
  let providerAscii = true;
  const systemProvider = system({ ascii: true });
  const imeProvider = provider({
    stateSource: {
      readAsciiMode() {
        return providerAscii;
      },
    },
    asciiModeSwitcher: {
      async switchAsciiMode(ascii) {
        providerSwitchCount++;
        providerAscii = ascii;
        return true;
      },
    },
  });
  const ime = new ImeGateway(imeProvider, systemProvider, logger);
  ime.refreshState();

  await ime.setAsciiMode(false);
  ime.refreshState();

  assert.equal(ime.getTrackedState().isAsciiMode, false);
  assert.equal(providerSwitchCount, 1);
  assert.equal(systemProvider.asciiSwitchCount, 0);
});

test('successful switch command does not replace observed input method state', async () => {
  let providerSwitchCount = 0;
  const imeProvider = provider({
    stateSource: {
      readAsciiMode: () => false,
    },
    asciiModeSwitcher: {
      async switchAsciiMode() {
        providerSwitchCount++;
        return true;
      },
    },
  });
  const ime = new ImeGateway(imeProvider, system({ ascii: true }), logger);
  ime.refreshState();

  await ime.setAsciiMode(true);

  assert.equal(ime.getTrackedState().isAsciiMode, false);
  assert.equal(providerSwitchCount, 1);
});

test('system change updates tracked input method state', async () => {
  const systemProvider = system({ ascii: true });
  const imeProvider = provider({
    asciiModeSwitcher: {
      async switchAsciiMode() {
        return true;
      },
    },
  });
  const ime = new ImeGateway(imeProvider, systemProvider, logger);
  ime.refreshState();
  await ime.setAsciiMode(true);

  systemProvider.setAsciiState(false);
  ime.refreshState();

  assert.equal(ime.getTrackedState().isAsciiMode, false);
});

test('system providers are selected through registry', () => {
  const registry = new SystemImeProviderRegistry();
  const windows = system();
  registry.register(SystemType.WINDOWS, () => windows);

  assert.equal(registry.create(SystemType.WINDOWS), windows);
  assert.throws(() => registry.create(SystemType.LINUX), /not found/);
  assert.equal(currentSystemType('win32'), SystemType.WINDOWS);
  assert.equal(currentSystemType('darwin'), SystemType.MACOS);
  assert.equal(currentSystemType('linux'), SystemType.LINUX);
});

test('declared unavailable state source suspends actions until it recovers', async () => {
  let available = false;
  let providerSwitchCount = 0;
  const imeProvider = provider({
    stateSource: {
      isAvailable: () => available,
      readState: () => ({ isAsciiMode: false, isComposing: false }),
    },
    asciiModeSwitcher: {
      async switchAsciiMode() {
        providerSwitchCount++;
        return true;
      },
    },
  });
  const ime = new ImeGateway(imeProvider, system({ ascii: true }), logger);

  ime.start();
  await ime.setAsciiMode(false);

  assert.equal(ime.isStateSourceAvailable(), false);
  assert.equal(providerSwitchCount, 0);

  available = true;
  ime.refreshState();
  await ime.setAsciiMode(true);

  assert.equal(ime.isStateSourceAvailable(), true);
  assert.equal(providerSwitchCount, 1);
});
